internal/handlers/handler_sign_webauthn.go

```diff
@@ -0,0 +1,188 @@
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	if err = ctx.SaveSession(userSession); err != nil {
+		ctx.Logger.Errorf(logFmtErrSessionSave, "assertion challenge", regulation.AuthTypeWebauthn, userSession.Username, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	if err = ctx.SetJSONBody(assertion); err != nil {
+		ctx.Logger.Errorf(logFmtErrWriteResponseBody, regulation.AuthTypeWebauthn, userSession.Username, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+}
+
+// SecondFactorWebauthnAssertionPOST handler completes the assertion ceremony after verifying the challenge.
+func SecondFactorWebauthnAssertionPOST(ctx *middlewares.AutheliaCtx) {
+	var (
+		err error
+		w   *webauthn.WebAuthn
+
+		requestBody signWebauthnRequestBody
+	)
+
+	if err = ctx.ParseBody(&requestBody); err != nil {
+		ctx.Logger.Errorf(logFmtErrParseRequestBody, regulation.AuthTypeWebauthn, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	userSession := ctx.GetSession()
+
+	if w, err = getWebauthn(ctx); err != nil {
+		ctx.Logger.Errorf("Unable to configire %s during assertion challenge for user '%s': %+v", regulation.AuthTypeWebauthn, userSession.Username, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	var (
+		assertionResponse *protocol.ParsedCredentialAssertionData
+		credential        *webauthn.Credential
+		user              *models.WebauthnUser
+	)
+
+	if assertionResponse, err = protocol.ParseCredentialRequestResponseBody(bytes.NewReader(ctx.PostBody())); err != nil {
+		ctx.Logger.Errorf("Unable to parse %s assertionfor user '%s': %+v", regulation.AuthTypeWebauthn, userSession.Username, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	if user, err = getWebAuthnUser(ctx, userSession); err != nil {
+		ctx.Logger.Errorf("Unable to load %s devices for assertion challenge for user '%s': %+v", regulation.AuthTypeWebauthn, userSession.Username, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	if credential, err = w.ValidateLogin(user, *userSession.Webauthn, assertionResponse); err != nil {
+		_ = markAuthenticationAttempt(ctx, false, nil, userSession.Username, regulation.AuthTypeWebauthn, err)
+
+		respondUnauthorized(ctx, messageMFAValidationFailed)
+
+		return
+	}
+
+	var found bool
+
+	for _, device := range user.Devices {
+		if bytes.Equal(device.KID.Bytes(), credential.ID) {
+			device.UpdateSignInInfo(w.Config, ctx.Clock.Now(), credential.Authenticator.SignCount)
+
+			found = true
+
+			if err = ctx.Providers.StorageProvider.UpdateWebauthnDeviceSignIn(ctx, device); err != nil {
+				ctx.Logger.Errorf("Unable to save %s device signin count for assertion challenge for user '%s': %+v", regulation.AuthTypeWebauthn, userSession.Username, err)
```

Comment:
I think I'm fine with the idea, it would immediately highlight issues if there were any in updating this information. At the time I didn't think of it as critical, but it could potentially hide bugs. Is that the primary motivation?

Reply:
I implemented this.